### PR TITLE
Add TestHarness, and assorted API breaking changes

### PR DESCRIPTION
**Pull request by SimonSapin**
This PR makes a number of changes, many of them breaking. Each commit should be self-contained, though. The highlights are:

* A new `TestHarness` API that replaces both `PluginTestHarness` (which is removed) and `PluggableRouterServiceBuilder` (which becomes private)
* The above allows a number of other items to be made private, as they are no longer needed outside of the crate or appear in public signatures
* The `services` module was also made private. It is replaced in the public API by a new `stages` module that contains sub-modules for each pipeline stage (`query_planner`, `execution`, etc) which in turn contain items like `Request`, `BoxService`, etc.
* Some other drive-by API changes, for polish towards 1.0

For more details, see `NEXT_CHANGELOG.md`, `apollo-router/src/test_harness.rs`, and `apollo-router/src/stages.rs`. Much of the other changes are very mechanical.

Many choices proposed in this PR are opinionated but everything is open to discussion.

It should probably have been multiple smaller PRs. I got a bit carried away, sorry. For review I suggest going commit-by-commit.

### CODE REVIEW

**Comment by SimonSapin** [comment]
For now CI will fail for (at least) two reasons:

* [ ] Some tracing test fails. This failure will go away once this PR is rebased on top of https://github.com/apollographql/router/pull/1463
* [ ] A warning in `apollo-router/src/plugins/telemetry/config.rs` because `MetricsCommon::delay_interval` is never used. The warning is new because the struct and field used to be public, but it’s legitimate. Should we remove this configuration since it does nothing?

**Comment by o0Ignition0o** [comment]
Not sure if it's in the scope of this pr, but it would be great if we tested our plugins against a yaml! configuration instead of 

```rust
        let config = serde_json::json!({
            "plugins": {
                "my_example.catch_query_planner_error": {
                    "enabled": true ,
                }
            }
        });
        apollo_router::TestHarness::builder()
            .configuration_json(config)
            .unwrap()
            .build()
            .await
            .unwrap();
    }
```

This would allow us to catch [yml related issues](https://github.com/apollographql/router/issues/1490#issue-1334282421) earlier

**Comment by o0Ignition0o** [comment]
```
                    let query_planner_response: Result<query_planner::Response, BoxError> =
                        query_planner_future.await;
```
@SimonSapin you spoiled me! I was expecting query_planner::Result to be available :D

**Comment by o0Ignition0o** [comment]
Any chance we can fix the requirement of an alias while we're at it ?

```rust
use apollo_router::layers::ServiceExt;
use tower::{ServiceExt as _};
```

**Comment by SimonSapin** [comment]
We could definitely add `Result` aliases!

Renaming `ServiceExt` too, but this PR is already big, let's do additional changes separately 😅

**Comment by SimonSapin** [comment]
We could definitely have a convenience builder method for `TestHarness` that accepts YAML, but it would have to be as a `&str` (so typically in a Rust string literal without syntax highlighting). `serde_yaml` doesn't have a macro like `serde_json::json!` because YAML syntax is too different from Rust syntax. `json!` actually takes Rust tokens as inputs, I'm sure there are edge cases where the syntax is not quite the same. It just looks close enough.

File: examples/cookies-to-headers/src/main.rs [hunk 1]
```diff
@@ -54,86 +48,83 @@ mod tests {
...
+                        let mock_data = response.context.get("mock_data").unwrap();
+                        let body = response.response.body_mut();
+                        let dummy = futures::stream::empty().boxed();
+                        let stream = std::mem::replace(body, dummy);
+                        *body = stream
+                            .map(move |mut resp| {
+                                resp.data = mock_data.clone();
+                                resp
+                            })
+                            .boxed();
```
Comment (o0Ignition0o) [inline examples/cookies-to-headers/src/main.rs:102]: is this why we need futures::StreamExt?

This feels a bit inconvenient, I m not sure I would assume plugin writers to be able to wield this, especially when learning rust :/

File: apollo-router/src/router.rs [hunk 2]
```diff
@@ -316,74 +332,142 @@ impl ShutdownKind {
...
+    ///
+    /// The server handle can be used in multiple ways.
+    /// As a [`Future`], it resolves to `Result<(), `[`ApolloRouterError`]`>`
+    /// either when the server has finished gracefully shutting down
+    /// or when it encounters a fatal error that prevents it from starting.
+    ///
+    /// If the handle is dropped before being awaited as a future,
+    /// a graceful shutdown is triggered.
+    /// In order to wait until shutdown finishes,
+    /// use the [`shutdown`][Self::shutdown] method instead.
```
Comment (o0Ignition0o) [inline apollo-router/src/router.rs:411]: wow 🤩

**Review by o0Ignition0o — APPROVED** [review]
LGTM!

I'd like to take a moment to acknowledge the HUGE amount of work you had to do to come up with such a big, and so well documented PR.

This one touches everything, and the same level of care and attention to detail has been given to each file, and each function.

I know I have suffered review fatigue while reading this, so I can't imagine the amount of work and concentration preparing such a changeset required. And I won't even mention the number of rebases and merge conflicts you had to go through.

Again kudos for the very nice modules and functions documentation. This is really great work!

**Comment by o0Ignition0o** [comment]
@SimonSapin I didn't see you enabled automerge, let's followup on the comment's I ve written next week!

File: examples/cookies-to-headers/src/main.rs (hunk 1 above)
Comment (SimonSapin) [reply examples/cookies-to-headers/src/main.rs:102]: Yes, this was an ugly way to get this test to compile. I’ve added a helper in https://github.com/apollographql/router/pull/1525
